python/ray/serve/api.py

```diff
@@ -193,24 +193,82 @@ def shutdown(self) -> None:
 
             self._shutdown = True
 
-    def _wait_for_goal(
-        self, goal_id: Optional[GoalId], timeout: Optional[float] = None
-    ) -> bool:
-        if goal_id is None:
-            return True
+    def _wait_for_deployments_shutdown(self, timeout_s: int = 60):
```

Comment:
I see your point as prototype type and it makes sense. For prod these functions can be moved to a separate component running in its own process.. we should also aim to make api.py simpler, it's too long now. How about something like:

user ->(targets_state) DB
watcher: Subscribes to DB and makes controller API calls (it's kinda like the reconciler on product)
controller: keeps a copy of its view of current deployment states and target state, provide mechanism to do rolling updates

Reply:
Yes for production use case you'd want a higher-level component monitoring this (e.g., a k8s operator or something). For the Python API, I think we want it here for convenience.

Reply:
The watcher part is orthogonal